File: scripts/Generate Parentheses/Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        
        generate(ans,"",n,n);
        
        return ans;
    }
    
    void generate(vector<string> &ans,string s,int open,int close)
    {
        if(open == 0 && close == 0)
        {
            ans.push_back(s);
            return ;
        }
        
        if(open > 0)
        {
            generate(ans,s+'(',open-1,close);
        }
        
        if(open < close)
        {
            generate(ans,s+')',open,close-1);
        }
    }
};

File: scripts/algorithms/G/Generate Parentheses/Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:

        # Initialize the result
        res = []

        # Recursively go through all possible combinations
        def add(open, close, partialRes):

            nonlocal res

            # If we have added opening and closing parentheses n times, we reaches a solution
            if open == close == n:
                res.append("".join(partialRes))
                return

            # Add a closing parenthesis to the partial result if we have at least 1 opening parenthesis
            if close < open:
                add(open, close + 1, partialRes + [")"])

            # Add an opening parenthesis to the partial result if we haven't added n parenthesis yet
            if open < n:
                add(open + 1, close, partialRes + ["("])

        add(0, 0, [])

        return res
